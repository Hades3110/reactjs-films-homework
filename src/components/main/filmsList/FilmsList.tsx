import { useState, useEffect } from "react"
import { ActionTypeInterface } from "../../../global"
import FilmItem from "./filmItem/FilmItem"
import { useSelector, useDispatch } from "react-redux"
import { changeFilmCategoriesAndCount } from "../../../redux/filmListCountAndCategories/action"
import {
    FilmInterface,
    getFilmlist,
    getGenres,
    GenreInterface,
    getSearchResult
} from "../../../services/filmService"
import uniqid from 'uniqid';
import './filmList.scss'
import spinner from '/public/assets/spinner.gif'
import { changeLoading } from "../../../redux/loading/action"

const FilmsList = () => {

    const [filmArr, setFilmArr] = useState<FilmInterface[]>([])
    const [genres, setGenres] = useState<GenreInterface[]>([])

    const dispatch = useDispatch()
    const pageCounter = useSelector((state: ActionTypeInterface) => state.pageCounter.count)
    const categories = useSelector((state: ActionTypeInterface) => state.pageCounter.categories)
    const sortType = useSelector((state: ActionTypeInterface) => state.sortType.type)
    const isLoaded = useSelector((state: ActionTypeInterface) => state.movieLoader.isLoaded)

    const search = useSelector((state: ActionTypeInterface) => state.mainBannerAndSearch.search)

    useEffect(() => {
        if (search.length > 0) {
            // const handle: ReturnType<typeof setTimeout> = setTimeout(() => {
            getSearchResult(search)
                .then(res => {
                    if (res.results.length > 0) {
                        setFilmArr(res.results)
                        dispatch(changeLoading(true))
                    } else {
                        console.log('e')
                    }

                })
            // }, 1000)

            // return () => clearTimeout(handle)
        } else {
            if (pageCounter > 1) {
                getFilmlist(categories, pageCounter)
                    .then(res => {
                        setFilmArr(() => ([...filmArr, ...res.results]))
                        dispatch(changeLoading(true))
                    })
            } else {
                getFilmlist(categories, pageCounter)
                    .then(res => {
                        setFilmArr(res.results)
                        dispatch(changeLoading(true))
                    })
            }
        }

    }, [categories, pageCounter, search])

    useEffect(() => {
        getGenres()
            .then(res => setGenres(res.genres))
    }, [])

    console.log(filmArr)
    return (
        <>
            {isLoaded ?
                <div className={sortType ? 'filmListRow' : 'filmListColumn'}>
                    {
                        filmArr.map((el: FilmInterface, i: number) => {
                            return (
                                <FilmItem
                                    sortType={sortType}
                                    key={uniqid()}
                                    title={el.title ?? 'Name not found'}
                                    image={el.poster_path}
                                    rate={el.vote_average}
                                    overview={el.overview ?? 'No description'}
                                    filmGenres={el.genre_ids}
                                    genres={genres}
                                />
                            )
                        })
                    }
                </div> : <img src={spinner} className="spinner" />}
            <div className="btn"
                onClick={() => {
                    dispatch(changeFilmCategoriesAndCount(pageCounter + 1, categories))
                }}>
                <button className="btn__more">Load More</button>
            </div>
        </>
    )
}

export default FilmsList
