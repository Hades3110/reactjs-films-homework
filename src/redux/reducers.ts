import { combineReducers } from "redux";
import filmListPageCountReducer from "./filmListCountAndCategories/reducer";
import filmListSortReducer from "./filmListSort/reducer";
import mainBannerReducer from "./mainBanner/reducer";

const rootReducer = combineReducers({
    pageCounter: filmListPageCountReducer,
    sortType: filmListSortReducer,
    mainBanner: mainBannerReducer
})

export default rootReducer
