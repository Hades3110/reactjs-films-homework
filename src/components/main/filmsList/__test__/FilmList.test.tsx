import { create } from "react-test-renderer";
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from '../../../../redux/reducers';
import FilmsList from '../FilmsList';

describe('FilmsList Component', () => {

  const store = createStore(rootReducer)

  it('FilmsList snapshot test', () => {
    const component = create(<Provider store={store}><FilmsList /></Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});