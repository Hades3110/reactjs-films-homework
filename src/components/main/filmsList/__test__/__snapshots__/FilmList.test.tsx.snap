// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmsList Component FilmsList snapshot test 1`] = `
<img
  alt="spinner"
  className="spinner"
  src="test-file-stub"
/>
`;
